### ISSUE TITLE
Difference in model implementation and Accuracy Discrepancy between TorchRec and Pytorch DLRMs 

### ISSUE THREAD

**Issue by allenfengjr** [meta-pytorch/torchrec#1001]
Hello,

I’m currently testing the two versions of DLRM, i.e., [one](https://github.com/facebookresearch/dlrm/blob/main/dlrm_s_pytorch.py) implemented in Pytorch and [one](https://github.com/facebookresearch/dlrm/blob/main/torchrec_dlrm/dlrm_main.py) implemented in TorchRec. I observed an accuracy discrepancy between these two versions during training. 

Specifically, I have been training the two DLRM versions/models on Criteo 1TB dataset; after one epoch, the testing accuracy of the TorchRec version converges to 96.6%. Please see the below figure (x-axis is # of iteration within one epoch).

<img width="375" alt="DLRM-accuracy" src="https://user-images.githubusercontent.com/5705572/215895721-c6f47b8f-fc16-4723-a4a5-6624da894873.png">

However, based on the [referenced figure](https://raw.githubusercontent.com/facebookresearch/dlrm/main/terabyte_0875_loss_accuracy_plots.png) from the Pytorch DLRM repo, the testing accuracy of the Pytorch version is supposed to be around 81%. 

Not only about the accuracy, I also checked the training/validation loss. The loss values I got from the TorchRec version (i.e., about 0.13, see below figure) are much lower than the [reference loss values](https://raw.githubusercontent.com/facebookresearch/dlrm/main/terabyte_0875_loss_accuracy_plots.png) (i.e., about 0.42) given by the Pytorch version. 

<img width="375" alt="DLRM-loss" src="https://user-images.githubusercontent.com/5705572/215895648-53deedcf-66f1-4c01-a2e4-64be7357f080.png">

I also verified that after one epoch the AUROC score of the TorchRec version can reach to 0.802, which is consistent with the number reported in the [TorchRec DLRM repo](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm). 

Below is my configuration. It is similar to the configuration provided in the [reference doc](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm#preliminary-training-results). The only difference is I used 16 GPUs from 4 nodes. I also checked the MD5 checksums of the training data files, which are exactly the same as the checksums in [checksum file](https://github.com/facebookresearch/dlrm/blob/main/torchrec_dlrm/md5sums_preprocessed_criteo_click_logs_dataset.txt). 

```
srun -n 16 python -u dlrm_main.py --epochs=1 \
--in_memory_binary_criteo_path="/N/scratch/haofeng/TB/processed" \
--num_embeddings_per_feature "45833188,36746,17245,7413,20243,3,7114,1441,62,29275261,1572176,345138,10,2209,11267,128,4,974,14,48937457,11316796,40094537,452104,12606,104,35" \
--embedding_dim 128 \
--batch_size 8192 \
--learning_rate 0.05 \
--over_arch_layer_sizes "1024,1024,512,256,1" \
--dense_arch_layer_sizes "512,256,128" \
--shuffle_batches \
--print_sharding_plan \
--mmap_mode

```

I also tried different learning rates, including 0.05, 0.5, 1, 7.5, 15, but their accuracies are still much higher than the reference accuracy around 81%.

I’m wondering if anyone possibly knows where the accuracy discrepancy could be coming from. Is there any reference of training/testing accuracy of the TorchRec DLRM version? What’s the main difference between these two models?

Thanks,
Hao

**Comment by colin2328** [comment]
Hi Hao,
If you are running those two examples, one is using DLRMv2 (i.e including a different DCNv2 arch, Adagrad optimizer) and the other is using DLRM (v1) .

**Comment by samiwilf** [comment]
@allenfengjr

Sub-sampled and non-sub-sampled results are not supposed to be identical. The [figure](https://raw.githubusercontent.com/facebookresearch/dlrm/main/terabyte_0875_loss_accuracy_plots.png) is titled "Terabyte Data (sub-sampled=0.875)".  Notice **sub-sampled=0.875**. Sub-sampling alters the prediction metrics. 

Global batch size is not matching.  Running 16 instances each with a local batch size of 8192 means the global batch size is 131,072. That size is larger than any global batch size shown in the [table](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm#preliminary-training-results).  One thing to watch out for is that --batch_size is the local batch_size in torchrec/dlrm_main.py, while --mini-batch-size is the global batch size in dlrm_s_pytorch.py.

Besides the two points above, your command to run dlrm_main.py looks good if you're attempting to run the same settings on both dlrm_s_pytorch.py and torchrec_dlrm/dlrm_main.py. The command would run an optimizer (SGD) that both scripts run by default and an interaction type (dot interaction) that both scripts run by default.
